### PR TITLE
debug_utils: Fix issues discovered by @krOoze

### PR DESCRIPTION
**Pull request by MarkY-LunarG**
Fixing a few items discovered by @krOoze with the EXT_debug_utils
extension spec language validation.  Namely in two issues:

Issue #789:
 The problem was the manual VUs should have been applied to
 the `vkSetDebugUtilsObjectNameEXT` command and not the
 general `VkDebugUtilsObjectNameInfoEXT` structure.

Issue #790:
 Here, we needed to remove the "noautovalidity" tag on
 the XML and add the "optional" tag to `pObjects` and
 `objectCount`.

### CODE REVIEW

File: chapters/VK_EXT_debug_utils.txt [hunk 1]
```diff
@@ -44,6 +44,14 @@ include::../api/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
 include::../validity/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
+.Valid Usage
+****
+  * pname:pNameInfo\->objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: If I change it, won't it break as soon as that change goes in?  Instead, if I leave it like this, it will just start working since that PR is likely going to go in soon, right?

File: chapters/VK_EXT_debug_utils.txt [hunk 2]
```diff
@@ -647,8 +644,20 @@ The parameters are passed on to the callback in addition to the
 pname:pUserData value that was defined at the time the messenger was
 registered.
 
+.Valid Usage
+****
+* If pname:pCallbackData\->objectCount is greater than 0, then for
+  each object in pname:pCallbackData\->objects, the value of
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: :+1:

File: chapters/VK_EXT_debug_utils.txt [hunk 3]
```diff
@@ -44,6 +44,14 @@ include::../api/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
 include::../validity/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
+.Valid Usage
+****
+  * pname:pNameInfo\->objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
+  * pname:pNameInfo\->objectHandle must: not be dlink:VK_NULL_HANDLE
+  * pname:pNameInfo\->objectHandle must: be a Vulkan object of the type associated with
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: :+1:

File: chapters/VK_EXT_debug_utils.txt [hunk 4]
```diff
@@ -44,6 +44,14 @@ include::../api/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
 include::../validity/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
+.Valid Usage
+****
+  * pname:pNameInfo\->objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
+  * pname:pNameInfo\->objectHandle must: not be dlink:VK_NULL_HANDLE
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: Could be.  This was how it was with the Debug Marker version.  I could remove it since a valid handle seems to imply non-NULL.  But I didn't want to assume that without looking through the spec first.  Where I see VK_NULL_HANDLE being used, it does seem to be an independent value from "a valid handle of type XXX".  So, if we do state that it must be a valid handle, I think that's good enough.  :+1:

File: chapters/VK_EXT_debug_utils.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (krOoze) [reply chapters/VK_EXT_debug_utils.txt]: That should be so. The PR is ancient though. After merge need to check for visible `pname`s anyway...

[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: Okay, I'll switch it, then.

File: chapters/VK_EXT_debug_utils.txt [hunk 5]
```diff
@@ -44,6 +44,14 @@ include::../api/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
 include::../validity/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
+.Valid Usage
+****
+  * pname:pNameInfo\->pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
+  * pname:pNameInfo\->pname:objectHandle must: be a Vulkan handle of the type associated with
+    pname:pNameInfo\->pname:objectType as defined in
+    <<debugging-object-types, VkObjectType and Vulkan Handle Relationship>>.
```
Comment (krOoze) [inline chapters/VK_EXT_debug_utils.txt]: "in the ... table"?
\+ No period after VU

File: chapters/VK_EXT_debug_utils.txt [hunk 6]
```diff
@@ -66,17 +74,6 @@ calling fname:vkSetDebugUtilsObjectNameEXT again with a new string.
...
 
-.Valid Usage
-****
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectType-01905]]
-    pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectHandle-01906]]
-    pname:objectHandle must: not be dlink:VK_NULL_HANDLE
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectHandle-01907]]
-    pname:objectHandle must: be a Vulkan object of the type associated with
-    pname:objectType as defined in <<debugging-object-types>>.
```
Comment (krOoze) [inline chapters/VK_EXT_debug_utils.txt]: Should still say "if not unknown type, must be valid handle". It might come up in some obscure case, e.g. `api_dump` layer after `unique_objects` layer -- i.e. the handle must be properly translatable without crashes.

File: chapters/VK_EXT_debug_utils.txt [hunk 7]
```diff
@@ -66,17 +74,6 @@ calling fname:vkSetDebugUtilsObjectNameEXT again with a new string.
 If pname:pObjectName is an empty string, then any previously set name is
 removed.
 
-.Valid Usage
-****
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectType-01905]]
-    pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectHandle-01906]]
-    pname:objectHandle must: not be dlink:VK_NULL_HANDLE
```
Comment (krOoze) [inline chapters/VK_EXT_debug_utils.txt]: Might still make sense to ban Unknown+Null. The app  an just pass 0 object count for that case.

Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: How about:

```
  * If pname:objectType is ename:VK_OBJECT_TYPE_UNKNOWN,
    pname:objectHandle must: not be dlink:VK_NULL_HANDLE
```

File: chapters/VK_EXT_debug_utils.txt (hunk 6 above)
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: How about this:

```
  * If pname:objectType is not ename:VK_OBJECT_TYPE_UNKNOWN,
    pname:objectHandle must: be a Vulkan handle of the type associated with
    pname:objectType as defined in the
    <<debugging-object-types, VkObjectType and Vulkan Handle Relationship>>
    table
```
That would allow me to remove the VU statement we just added at `vkSetDebugUtilsObjectNameEXT` since that covers this case.  The only VU statement left for `vkSetDebugUtilsObjectNameEXT` would be this:

```
  * pname:pNameInfo\->pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
```

Since we don't want UNKNOWN there.

File: chapters/VK_EXT_debug_utils.txt (hunk 5 above)
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: :+1:

File: chapters/VK_EXT_debug_utils.txt (hunk 6 above)
Comment (krOoze) [reply chapters/VK_EXT_debug_utils.txt]: Wait I am starting to get lost. Let's recapitulate.

There are two users of `VkDebugUtilsObjectNameInfoEXT`:  `vkSetDebugUtilsObjectNameEXT` and `VkDebugUtilsMessengerCallbackDataEXT` (which in turn is used only by `vkSubmitDebugUtilsMessageEXT`; and by user's callback but that does not count towards VU).

For `vkSetDebugUtilsObjectNameEXT` we always want the and only valid type **and** valid handle.

For `vkSubmitDebugUtilsMessageEXT` I think we agreed it should be valid to provide:
- valid type + valid handle
- valid type + null handle

(
The invalid usage for `vkSubmitDebugUtilsMessageEXT` therefore should be:
- unknown type
- valid type + invalid non-null handle

)

That about right?

<hr />

So if we do GCD, then `VkDebugUtilsObjectNameInfoEXT` can have the
"must not be `UNKNOWN_TYPE`" VU,

The `vkSetDebugUtilsObjectNameEXT` then needs
"`object` must be valid `objectType` handle" VU.

And the `vkSubmitDebugUtilsMessageEXT` would need
"If `object` is not NULL, `object` must be valid `objectType` handle" VU.

(Which alternatively could be moved to `VkDebugUtilsObjectNameInfoEXT`, and then `vkSetDebugUtilsObjectNameEXT` would not need to note the handle type: "`object` must be a valid Vulkan handle")

File: chapters/VK_EXT_debug_utils.txt (hunk 7 above)
Comment (krOoze) [reply chapters/VK_EXT_debug_utils.txt]: Actually that would ban exactly the case we wanted to allow (i.e. allow NULL_HANDLE object from `vkSubmitDebugUtilsMessageEXT`)?

Comment (krOoze) [reply chapters/VK_EXT_debug_utils.txt]: I think we agreed that we do not want `VK_OBJECT_TYPE_UNKNOWN` cases with valid handle, or did you reconsider?

Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: No, I'm still leaning towards allowing it because the validation layers could return it.  If it's in a valid usage statement on the object, people might think the layers are behaving incorrectly.

In the case of an extension not supported by the layers, that adds a new type, it may call a supported command with the new type value which won't be known resulting in an error message with `UNKNOWN`.

File: chapters/VK_EXT_debug_utils.txt (hunk 6 above)
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: We might have `VK_OBJECT_TYPE_UNKNOWN` type for `VkDebugUtilsObjectNameInfoEXT` when it's returned in the callback data if the layers fail to determine a handle in some cases.  This can be the case when an extension adds a type and isn't implemented in the layers (but is in the driver). 

So, I don't think "must not be UNKNOWN_TYPE" VU, belongs on `VkDebugUtilsObjectNameInfoEXT`.  Instead, I think it needs to be called out during the two commands.

This is the way I see it:

`VkDebugUtilsObjectNameInfoEXT`
  * If type == unknown, handle must not be NULL

`vkSubmitDebugUtilsMessageEXT`
  * For each object in pCallbackData->pObjects, type must: not be Unknown
  * For each object in pCallbackData->pObjects, if handle not NULL must be a valid handle

`vkSetDebugUtilsObjectNameEXT`
 * pNameInfo->objectType must: not be Unknown
 * pNameInfo->objectHandle must: be a valid handle

Comment (krOoze) [reply chapters/VK_EXT_debug_utils.txt]: Hmm, I never considered VUs to apply to stuff the Implementation returns (or calls calbacks with)...
I think returned values (and parameters from Implementation to the Applications callback) needs its own restrictions clerly addressed to the Implementers.

Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: `vkSubmitDebugUtilsMessageEXT` is stricter because the application should know everything about the objects that underlying layers/ICDs may not.  That way, we force them to pass in useful data.

Comment (krOoze) [reply chapters/VK_EXT_debug_utils.txt]: Then again, why should we care what the app passes to its own callback? Especially if the callback must already be able to handle the unrestricted cases.

Besides "application" may be just another layer. layer 2 (some library) may not know much more than the layers and ICD does.

And it is also weird to me that the layers/ICDs would not know. Either they do not know a thing is a handle\object in the first place, or they do know the type, right?

File: chapters/VK_EXT_debug_utils.txt [hunk 8]
```diff
@@ -44,6 +44,12 @@ include::../api/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
 include::../validity/protos/vkSetDebugUtilsObjectNameEXT.txt[]
 
+.Valid Usage
+****
+  * pname:pNameInfo\->pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
+  * pname:pNameInfo\->pname:objectHandle must: not be dlink:VK_NULL_HANDLE
+****
```
Comment (krOoze) [inline chapters/VK_EXT_debug_utils.txt]: This VU block is in the wrong place. Should be above Implicit VU.

File: chapters/VK_EXT_debug_utils.txt [hunk 9]
```diff
@@ -647,8 +654,15 @@ The parameters are passed on to the callback in addition to the
 pname:pUserData value that was defined at the time the messenger was
 registered.
 
+.Valid Usage
+****
+  * For each object in pname:pCallbackData\->pname:pObjects,
+    pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN.
```
Comment (krOoze) [inline chapters/VK_EXT_debug_utils.txt]: maybe
> pname:objectType member of each element of pname:pCallbackData\\->pname:pObjects must: not be ename:VK_OBJECT_TYPE_UNKNOWN

?
Also, no period for consistency.

Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: :+1:

File: chapters/VK_EXT_debug_utils.txt [hunk 10]
```diff
@@ -68,13 +74,14 @@ removed.
...
-    pname:objectType must: not be ename:VK_OBJECT_TYPE_UNKNOWN
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectHandle-01906]]
+  * If pname:objectType is ename:VK_OBJECT_TYPE_UNKNOWN,
     pname:objectHandle must: not be dlink:VK_NULL_HANDLE
-  * [[VUID-VkDebugUtilsObjectNameInfoEXT-objectHandle-01907]]
-    pname:objectHandle must: be a Vulkan object of the type associated with
-    pname:objectType as defined in <<debugging-object-types>>.
+  * If pname:objectType is not ename:VK_OBJECT_TYPE_UNKNOWN,
+    pname:objectHandle must: be dlink:VK_NULL_HANDLE or
+    a Vulkan handle of the type associated with pname:objectType
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: :+1:

File: chapters/VK_EXT_debug_utils.txt (hunk 8 above)
Comment (MarkY-LunarG) [reply chapters/VK_EXT_debug_utils.txt]: :+1:

File: xml/vk.xml [hunk 11]
```diff
@@ -6213,20 +6213,20 @@ server.
         </command>
         <command successcodes="VK_SUCCESS" errorcodes="VK_ERROR_OUT_OF_HOST_MEMORY">
             <proto><type>VkResult</type> <name>vkCreateDebugUtilsMessengerEXT</name></proto>
-            <param><type>VkInstance</type> <name>instance</name></param>
+            <param externsync="true"><type>VkInstance</type> <name>instance</name></param>
```
Comment (krOoze) [inline xml/vk.xml]: I think we would also need to add implicit externsync for every child of Instance. That seems unsustainable. There's also a subtle case that we probably only care about the dispatch parameters.

Maybe just add it as a freeform paragraph below `vkCreateDebugUtilsMessengerEXT`, e.g.:

> The application **must** ensure that `vkCreateDebugUtilsMessengerEXT` duration does not execute in parallel with duration of any Vulkan command that has been called with `instance` or child of `instance` as the dispatchable argument.

File: xml/vk.xml [hunk 12]
```diff
@@ -6213,20 +6213,20 @@ server.
...
             <proto><type>void</type> <name>vkDestroyDebugUtilsMessengerEXT</name></proto>
-            <param><type>VkInstance</type> <name>instance</name></param>
+            <param externsync="true"><type>VkInstance</type> <name>instance</name></param>
             <param externsync="true"><type>VkDebugUtilsMessengerEXT</type> <name>messenger</name></param>
             <param optional="true">const <type>VkAllocationCallbacks</type>* <name>pAllocator</name></param>
         </command>
         <command>
             <proto><type>void</type> <name>vkSubmitDebugUtilsMessageEXT</name></proto>
-            <param><type>VkInstance</type> <name>instance</name></param>
+            <param externsync="true"><type>VkInstance</type> <name>instance</name></param>
```
Comment (krOoze) [inline xml/vk.xml]: Not sure this needs `externsync`. The Impl probably does not need to modify instance here (only read the callback pointer, and get the filter parameters).

File: xml/vk.xml (hunk 11 above)
Comment (MarkY-LunarG) [reply xml/vk.xml]: I think you're statement is a little heavy handed.  "Child of `instance`" is basically any object.  I don't think it will impact anything if we just stated `instance`.

Comment (krOoze) [reply xml/vk.xml]: > Child of instance" is basically any object.

Well that's intentional, no? Debug messengers affect any and all objects. E.g. command called with `device`, where that `device` is child of that `instance`.

For comparison wording of the same situation in `vkGetInstanceProcAddr`:

> The returned function pointer **must** only be called with a dispatchable object (the first parameter) that is `instance` or a child of `instance`, e.g. `VkInstance`, `VkPhysicalDevice`, `VkDevice`, `VkQueue`, or `VkCommandBuffer`.

Comment (MarkY-LunarG) [reply xml/vk.xml]: Ok, I don't think it's going to matter that much, but better safe than sorry.  I did clean up your wording a little, too.

**Comment by krOoze** [comment]
BTW, I have gone ahead and posted https://github.com/KhronosGroup/Vulkan-Loader/issues/86. May I assume making the `objectCount` `optional` is intentional?

File: chapters/debugging.txt [hunk 13]
```diff
@@ -68,9 +68,6 @@ ifdef::VK_NVX_device_generated_commands[]
 | ename:VK_OBJECT_TYPE_OBJECT_TABLE_NVX | slink:VkObjectTableNVX
 | ename:VK_OBJECT_TYPE_INDIRECT_COMMANDS_LAYOUT_NVX | slink:VkIndirectCommandsLayoutNVX
 endif::VK_NVX_device_generated_commands[]
-ifdef::VK_EXT_debug_utils[]
-| ename:VK_OBJECT_TYPE_DEBUG_UTILS_MESSENGER_EXT | slink:VkDebugUtilsMessengerEXT
-endif::VK_EXT_debug_utils[]
```
Comment (krOoze) [inline chapters/debugging.txt]: Why is this removed?

File: chapters/VK_EXT_debug_utils.txt [hunk 14]
```diff
@@ -378,6 +385,14 @@ include::../api/protos/vkCreateDebugUtilsMessengerEXT.txt[]
 
 include::../validity/protos/vkCreateDebugUtilsMessengerEXT.txt[]
 
+[NOTE]
+.Note
+====
```
Comment (krOoze) [inline chapters/VK_EXT_debug_utils.txt]: Why as a Note? Specifying language should be outside a note IMO. Notes should only reiterate information that is already there but easy to miss from a formal description.

File: chapters/debugging.txt (hunk 13 above)
Comment (MarkY-LunarG) [reply chapters/debugging.txt]: You told me to remove it since you were commiting this change in another submission.

Comment (krOoze) [reply chapters/debugging.txt]: Right. Weird it shows in the diff then.
The change was merged, so I think you should rebase, just to be safe.

Comment (MarkY-LunarG) [reply chapters/debugging.txt]: Weird, looks like it got messed up even after a rebase.  I fixed it.
